craft_application/application.py

```diff
@@ -161,6 +170,23 @@ def __init__(
         # The kind of sessions that the fetch-service service should create
         self._fetch_service_policy = "strict"
 
+        # Load the plugins after initialization but before run
+        self._load_plugins()
+
+    @final
+    def _load_plugins(self) -> None:
+        """Load application plugins."""
+        # https://packaging.python.org/en/latest/specifications/entry-points/#data-model
+        for plugin_entry_point in metadata.entry_points(
+            group="craft_application_plugins.application"
+        ):
+            craft_cli.emit.debug(f"Loading app plugin {plugin_entry_point.name}")
+            try:
+                app_plugin: CraftApplicationPluginProtocol = plugin_entry_point.load()
+                app_plugin.configure(self)
+            except Exception as e:  # noqa: BLE001
+                craft_cli.emit.debug(f"Plugin failed: {e}")
```

Comment:
This we might want to actually warn the user about:

```suggestion
                craft_cli.emit.progress(f"Failed to load plugin {plugin_entry_point.name}", permanent=True)
                craft_cli.emit.debug(repr(e))
```